parquet/encoding.py

```diff
@@ -1,203 +1,216 @@
-    return conv(fo)
-
-
-def read_unsigned_var_int(fo):
+def read_plain(raw_bytes, type_, count, width=0):
+    if type_ in DECODE_TYPEMAP:
+        dtype = DECODE_TYPEMAP[type_]
+        return np.frombuffer(memoryview(raw_bytes), dtype=dtype, count=count)
+    if type_ == parquet_thrift.Type.FIXED_LEN_BYTE_ARRAY:
+        dtype = np.dtype('S%i' % width)
+        return np.frombuffer(memoryview(raw_bytes), dtype=dtype, count=count)
+    if type_ == parquet_thrift.Type.BOOLEAN:
+        return read_plain_boolean(raw_bytes, count)
+    # variable byte arrays (rare)
+    file_obj = io.BytesIO(raw_bytes)
+    return np.array([file_obj.read(struct.unpack('<i', file_obj.read(4))[0])
+                     for _ in range(count)], dtype="O")
+
+
+@numba.jit(nogil=True)
+def read_unsigned_var_int(file_obj):
+    """Read a value using the unsigned, variable int encoding.
+    file-obj is a NumpyIO of bytes; avoids struct to allow numba-jit
+    """
     result = 0
     shift = 0
     while True:
-        byte = struct.unpack("<B", fo.read(1))[0]
+        byte = file_obj.read_byte()
         result |= ((byte & 0x7F) << shift)
         if (byte & 0x80) == 0:
             break
         shift += 7
     return result
 
 
-def byte_width(bit_width):
-    "Returns the byte width for the given bit_width"
-    return (bit_width + 7) / 8
-
-
-def read_rle(fo, header, bit_width):
-    """Read a run-length encoded run from the given fo with the given header
-    and bit_width.
+@numba.njit(nogil=True)
+def read_rle(file_obj, header, bit_width, o):
+    """Read a run-length encoded run from the given fo with the given header and bit_width.
 
     The count is determined from the header and the width is used to grab the
     value that's repeated. Yields the value repeated count times.
     """
     count = header >> 1
-    zero_data = b"\x00\x00\x00\x00"
-    data = b""
-    width = byte_width(bit_width)
-    if width >= 1:
-        data += fo.read(1)
-    if width >= 2:
-        data += fo.read(1)
-    if width >= 3:
-        data += fo.read(1)
-    if width == 4:
-        data += fo.read(1)
-    data = data + zero_data[len(data):]
-    value = struct.unpack("<i", data)[0]
-    for i in range(count):
-        yield value
+    width = (bit_width + 7) // 8
+    zero = np.zeros(4, dtype=np.int8)
+    data = file_obj.read(width)
+    zero[:len(data)] = data
+    value = zero.view(np.int32)
+    o.write_many(value, count)
 
 
+@numba.njit(nogil=True)
 def width_from_max_int(value):
-    """Converts the value specified to a bit_width."""
-    return int(math.ceil(math.log(value + 1, 2)))
+    """Convert the value specified to a bit_width."""
+    for i in range(1, 64):
+        value >>= 1
+        if value == 0:
+            return i
+    return 1
 
 
+@numba.njit(nogil=True)
 def _mask_for_bits(i):
-    """Helper function for read_bitpacked to generage a mask to grab i bits."""
+    """Generate a mask to grab `i` bits from an int value."""
     return (1 << i) - 1
```

Comment:
Yes, I believe LLVM considers inlining functions, but I don't know how to check whether it actually does so.